File: scripts/build-entity-ssg-fixtures.mjs
#!/usr/bin/env node
/**
 * P5-E.9F.2 — Fixture-based Entity SSG generator.
 * Input: qa/fixtures/entity-ssg-fixtures.json (default)
 *        qa/fixtures/p5-entity-ssg-fixtures.json (--legacy)
 * Output: wiki/post/<canonical_slug>/index.html
 *         qa/entity-ssg-sitemap.fixture.xml
 *         wiki/post/_ssg-not-found/index.html
 * No network. No DB. No .env.
 */

import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(__dirname, "..");
const DEFAULT_FIXTURE = path.join(ROOT, "qa", "fixtures", "entity-ssg-fixtures.json");
const LEGACY_FIXTURE = path.join(ROOT, "qa", "fixtures", "p5-entity-ssg-fixtures.json");
const OUTPUT_BASE = path.join(ROOT, "wiki", "post");
const SITEMAP_OUT = path.join(ROOT, "qa", "entity-ssg-sitemap.fixture.xml");
const NOT_FOUND_OUT = path.join(OUTPUT_BASE, "_ssg-not-found", "index.html");
const CANONICAL_BASE = "https://boundlore.com";

const REQUIRED_FIELDS = [
  "canonical_slug",
  "title",
  "category",
  "entity_domain",
  "entity_subtype",
  "excerpt",
  "body_html_sanitized",
  "updated_at",
  "published_at",
  "content_origin",
  "status",
  "canonical_url",
  "breadcrumbs",
];

const ALLOWED_CONTENT_ORIGINS = new Set(["ssg_fixture", "prototype_fixture"]);
const SLUG_RE = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;
const SLUG_BLOCKLIST_RE = /^(qa|test|p5|fixture|draft|pending|deleted|contribution)/;
const FORBIDDEN_HTML_RE =
  /BLMETA|search_text|search_vector|service_role|SUPABASE_SERVICE_ROLE|@[a-z0-9.-]+\.[a-z]{2,}|\b[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\b/i;
const DEFAULT_IMAGE = CANONICAL_BASE + "/public/images/social-preview.png";
const GENERATOR_VERSION = "p5-e9f2";

function fail(message) {
  console.error("[build-entity-ssg-fixtures] ERROR:", message);
  process.exit(1);
}

function escapeHtml(value) {
  return String(value ?? "")
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;");
}

function escapeAttr(value) {
  return escapeHtml(value);
}

function truncate(text, max) {
  const s = String(text || "").trim();
  if (s.length <= max) return s;
  return s.slice(0, max - 1).trimEnd() + "…";
}

function validateSlug(slug) {
  if (!slug || typeof slug !== "string") return "canonical_slug missing";
  if (slug.includes("..") || slug.includes("/") || slug.includes("\\")) return "unsafe slug path";
  if (slug !== slug.toLowerCase()) return "slug must be lowercase";
  if (!SLUG_RE.test(slug)) return "slug must match [a-z0-9-]";
  if (SLUG_BLOCKLIST_RE.test(slug)) return "slug blocked (qa/test/fixture/draft/pending/deleted)";
  if (slug.startsWith("_")) return "slug must not start with underscore except reserved _ssg-not-found";
  return null;
}

function validateNoLeak(text, label) {
  const s = String(text || "");
  if (FORBIDDEN_HTML_RE.test(s)) return label + " contains forbidden leak pattern";
  if (/\b(draft|pending|deleted)\b/i.test(s) && /\bstatus\b/i.test(s)) {
    return label + " contains draft/pending/deleted status hint";
  }
  return null;
}

function validateBodyHtml(html) {
  const body = String(html || "");
  if (!body.trim()) return "body_html_sanitized empty";
  if (/<script\b/i.test(body)) return "script tag in body";
  if (/\bon\w+\s*=/i.test(body)) return "inline event handler in body";
  if (/javascript:/i.test(body)) return "javascript: URL in body";
  if (/<!--\s*blmeta/i.test(body)) return "BLMETA in body";
  return validateNoLeak(body, "body_html_sanitized");
}

function validateEntity(entity, index) {
  const label = "entities[" + index + "]";

  for (const field of REQUIRED_FIELDS) {
    if (entity[field] === undefined || entity[field] === null || entity[field] === "") {
      return label + " missing required field: " + field;
    }
  }

  if (entity.status !== "published") {
    return label + " status must be published";
  }

  if (!ALLOWED_CONTENT_ORIGINS.has(entity.content_origin)) {
    return label + " content_origin not allowed";
  }

  const slugErr = validateSlug(entity.canonical_slug);
  if (slugErr) return label + " " + slugErr;

  const expectedCanonical = CANONICAL_BASE + "/wiki/post/" + entity.canonical_slug + "/";
  if (entity.canonical_url !== expectedCanonical) {
    return label + " canonical_url mismatch";
  }

  if (!Array.isArray(entity.breadcrumbs) || entity.breadcrumbs.length < 2) {
    return label + " breadcrumbs invalid";
  }

  const bodyErr = validateBodyHtml(entity.body_html_sanitized);
  if (bodyErr) return label + " " + bodyErr;

  const excerptErr = validateNoLeak(entity.excerpt, "excerpt");
  if (excerptErr) return label + " " + excerptErr;

  const titleErr = validateNoLeak(entity.title, "title");
  if (titleErr) return label + " " + titleErr;

  return null;
}

function hubPathForCategory(category) {
  const map = {
    creatures: "/wiki/creatures/",
    items: "/wiki/items/",
    biomes: "/wiki/biomes/",
    locations: "/wiki/locations/",
    resources: "/wiki/resources/",
    guides: "/wiki/guides/",
    guilds: "/wiki/guilds/",
    lore: "/wiki/lore/",
  };
  return map[String(category || "").toLowerCase()] || "/wiki/browse/";
}

function buildBreadcrumbHtml(breadcrumbs) {
  const items = breadcrumbs.map(function (crumb, idx) {
    const name = escapeHtml(crumb.name);
    const isLast = idx === breadcrumbs.length - 1;
    if (isLast) return name;
    const href = escapeAttr(crumb.url || "#");
    return '<a href="' + href + '">' + name + "</a>";
  });
  return items.join(" &rsaquo; ");
}

function buildBadges(entity) {
  const badges = [
    "Category: " + entity.category,
    "Domain: " + entity.entity_domain,
    "Subtype: " + entity.entity_subtype,
  ];
  if (entity.evidence_tier) badges.push("Evidence: " + entity.evidence_tier);
  if (entity.completeness) badges.push("Completeness: " + entity.completeness);
  return badges
    .map(function (text) {
      return '        <span class="bl-ssg-badge">' + escapeHtml(text) + "</span>";
    })
    .join("\n");
}

function buildRelatedHtml(entity) {
  const related = Array.isArray(entity.relations_summary) ? entity.relations_summary : [];
  if (related.length === 0) return "";
  const links = related
    .map(function (rel) {
      const parts = String(rel).split(":");
      const slug = parts.length > 1 ? parts.slice(1).join(":") : parts[0];
      const label = escapeHtml(slug.replace(/-/g, " "));
      const href = "/wiki/post/" + encodeURIComponent(slug) + "/";
      return '          <li><a href="' + href + '">' + label + "</a></li>";
    })
    .join("\n");
  return (
    '      <section class="bl-ssg-related" aria-label="Related entries">\n' +
    "        <h2>Related</h2>\n" +
    "        <ul>\n" +
    links +
    "\n        </ul>\n" +
    "      </section>\n"
  );
}

function buildCreativeWorkJson(entity) {
  const image = entity.image_url || DEFAULT_IMAGE;
  return {
    "@context": "https://schema.org",
    "@type": "CreativeWork",
    name: entity.title,
    description: truncate(entity.excerpt, 160),
    url: entity.canonical_url,
    datePublished: entity.published_at,
    dateModified: entity.updated_at,
    image: image,
    keywords: [entity.category, entity.entity_domain, entity.entity_subtype].join(", "),
    isPartOf: { "@type": "WebSite", name: "BoundLore", url: CANONICAL_BASE + "/" },
  };
}

function buildBreadcrumbJson(entity) {
  return {
    "@context": "https://schema.org",
    "@type": "BreadcrumbList",
    itemListElement: entity.breadcrumbs.map(function (crumb, idx) {
      return {
        "@type": "ListItem",
        position: idx + 1,
        name: crumb.name,
        item: crumb.url,
      };
    }),
  };
}

function renderPage(entity) {
  const slug = entity.canonical_slug;
  const title = entity.title;
  const metaDesc = escapeAttr(entity.excerpt);
  const ogDesc = escapeAttr(truncate(entity.excerpt, 120));
  const twDesc = escapeAttr(truncate(entity.excerpt, 100));
  const canonical = escapeAttr(entity.canonical_url);
  const pageTitle = escapeHtml(title + " – BoundLore Wiki");
  const ogTitle = escapeAttr(title + " – BoundLore Wiki");
  const image = escapeAttr(entity.image_url || DEFAULT_IMAGE);
  const hubPath = hubPathForCategory(entity.category);
  const breadcrumbHtml = buildBreadcrumbHtml(entity.breadcrumbs);
  const badgesHtml = buildBadges(entity);
  const relatedHtml = buildRelatedHtml(entity);
  const creativeWork = JSON.stringify(buildCreativeWorkJson(entity), null, 2);
  const breadcrumbLd = JSON.stringify(buildBreadcrumbJson(entity), null, 2);

  return (
    "<!-- generated by scripts/build-entity-ssg-fixtures.mjs " +
    GENERATOR_VERSION +
    " -->\n" +
    '<!DOCTYPE html>\n' +
    '<html lang="en" data-bl-ssg="1" data-bl-canonical-slug="' +
    escapeAttr(slug) +
    '" data-bl-ssg-source="fixture-generator">\n' +
    "<head>\n" +
    '  <meta charset="UTF-8" />\n' +
    '  <meta name="viewport" content="width=device-width, initial-scale=1.0" />\n' +
    '  <meta name="description" content="' +
    metaDesc +
    '" />\n' +
    '  <meta name="robots" content="noindex, follow" />\n' +
    '  <link rel="canonical" href="' +
    canonical +
    '" />\n' +
    '  <meta property="og:title" content="' +
    ogTitle +
    '" />\n' +
    '  <meta property="og:description" content="' +
    ogDesc +
    '" />\n' +
    '  <meta property="og:type" content="article" />\n' +
    '  <meta property="og:url" content="' +
    canonical +
    '" />\n' +
    '  <meta property="og:image" content="' +
    image +
    '" />\n' +
    '  <meta name="twitter:card" content="summary_large_image" />\n' +
    '  <meta name="twitter:title" content="' +
    ogTitle +
    '" />\n' +
    '  <meta name="twitter:description" content="' +
    twDesc +
    '" />\n' +
    '  <meta name="twitter:image" content="' +
    image +
    '" />\n' +
    "  <title>" +
    pageTitle +
    "</title>\n" +
    '  <link rel="icon" type="image/jpeg" href="/public/images/icon.jpg" />\n' +
    '  <link rel="stylesheet" href="../../../css/style.css" />\n' +
    '  <link rel="preconnect" href="https://fonts.googleapis.com" />\n' +
    '  <link href="https://fonts.googleapis.com/css2?family=Cinzel:wght@400;700;900&family=Inter:wght@300;400;600&display=swap" rel="stylesheet" />\n' +
    '  <script src="/js/error-reporter.js?v=p5-e9c2"></script>\n' +
    '  <script type="application/ld+json">\n' +
    creativeWork +
    "\n  </script>\n" +
    '  <script type="application/ld+json">\n' +
    breadcrumbLd +
    "\n  </script>\n" +
    "  <style>\n" +
    "    .bl-ssg-badges { display: flex; flex-wrap: wrap; gap: 8px; margin: 12px 0 16px; }\n" +
    "    .bl-ssg-badge { padding: 4px 10px; border-radius: 999px; font-size: 0.75rem; border: 1px solid rgba(255,255,255,0.2); background: rgba(255,255,255,0.06); }\n" +
    "    .bl-ssg-excerpt { font-size: 1.05rem; color: #dde1ef; margin-bottom: 20px; line-height: 1.55; }\n" +
    "    .bl-ssg-body { line-height: 1.6; color: #e8ebf5; }\n" +
    "    .bl-ssg-related { margin-top: 28px; }\n" +
    "    .bl-ssg-related h2 { font-size: 1.1rem; margin-bottom: 10px; }\n" +
    "    .bl-ssg-related ul { padding-left: 1.2rem; }\n" +
    "    .bl-ssg-hydrate-slot { margin-top: 28px; padding: 16px; border: 1px dashed rgba(255,255,255,0.15); border-radius: 8px; color: #9aa8bc; font-size: 0.85rem; }\n" +
    "  </style>\n" +
    "</head>\n" +
    "<body>\n" +
    '  <nav class="navbar" id="navbar">\n' +
    '    <div class="nav-container">\n' +
    '      <a href="/" class="nav-logo">\n' +
    '        <img src="/public/images/icon.jpg" alt="BoundLore" class="nav-icon" />\n' +
    '        <span class="logo-text">Bound<span class="accent">Lore</span></span>\n' +
    "      </a>\n" +
    '      <ul class="nav-links" id="navLinks">\n' +
    '        <li><a href="/">Home</a></li>\n' +
    '        <li><a href="/wiki/creatures/">Creatures</a></li>\n' +
    '        <li><a href="/wiki/biomes/">Biomes</a></li>\n' +
    '        <li><a href="/wiki/items/">Items</a></li>\n' +
    '        <li><a href="/wiki/browse/">Browse</a></li>\n' +
    "      </ul>\n" +
    "    </div>\n" +
    "  </nav>\n" +
    "\n" +
    '  <div class="article-layout bl-post-layout" id="blSsgHydrateRoot" data-bl-ssg-hydrate="1" data-bl-ssg-source="fixture-generator" data-bl-canonical-slug="' +
    escapeAttr(slug) +
    '">\n' +
    '    <article class="article-content bl-post-main">\n' +
    '      <p class="breadcrumb">' +
    breadcrumbHtml +
    "</p>\n" +
    "\n" +
    "      <h1>" +
    escapeHtml(title) +
    "</h1>\n" +
    "\n" +
    '      <div class="bl-ssg-badges">\n' +
    badgesHtml +
    "\n" +
    "      </div>\n" +
    "\n" +
    '      <p class="bl-ssg-excerpt">' +
    escapeHtml(entity.excerpt) +
    "</p>\n" +
    "\n" +
    '      <div class="bl-ssg-body">\n' +
    entity.body_html_sanitized +
    "\n" +
    "      </div>\n" +
    "\n" +
    relatedHtml +
    "\n" +
    '      <div class="bl-ssg-hydrate-slot" id="blSsgHydrateRelated" aria-label="CSR hydration placeholder">\n' +
    "        Comments and reactions may load here after hydration. SEO core content is fully static above.\n" +
    "      </div>\n" +
    "    </article>\n" +
    "  </div>\n" +
    "\n" +
    '  <footer class="footer"><div class="footer-bottom"><p>&copy; 2026 BoundLore Community &middot; <a href="/wiki/imprint/">Imprint</a> &middot; <a href="/wiki/privacy/">Privacy Policy</a></p></div></footer>\n' +
    "</body>\n" +
    "</html>\n"
  );
}

function renderNotFoundPage() {
  const canonical = CANONICAL_BASE + "/wiki/post/_ssg-not-found/";
  return (
    "<!-- generated by scripts/build-entity-ssg-fixtures.mjs " +
    GENERATOR_VERSION +
    " -->\n" +
    '<!DOCTYPE html>\n' +
    '<html lang="en" data-bl-ssg="1" data-bl-ssg-not-found="1">\n' +
    "<head>\n" +
    '  <meta charset="UTF-8" />\n' +
    '  <meta name="viewport" content="width=device-width, initial-scale=1.0" />\n' +
    '  <meta name="description" content="The requested wiki entry was not found." />\n' +
    '  <meta name="robots" content="noindex, nofollow" />\n' +
    '  <link rel="canonical" href="' +
    canonical +
    '" />\n' +
    "  <title>Entry Not Found – BoundLore Wiki</title>\n" +
    '  <link rel="stylesheet" href="../../../css/style.css" />\n' +
    "</head>\n" +
    "<body>\n" +
    '  <main class="article-content" style="max-width:720px;margin:80px auto;padding:24px;">\n' +
    "    <h1>Entry Not Found</h1>\n" +
    "    <p>The requested wiki entry is not available as static SSG output.</p>\n" +
    '    <p><a href="/wiki/browse/">Return to Browse</a></p>\n' +
    "  </main>\n" +
    "</body>\n" +
    "</html>\n"
  );
}

function buildSitemapXml(entities) {
  const sorted = entities.slice().sort(function (a, b) {
    return a.canonical_slug.localeCompare(b.canonical_slug);
  });
  const urls = sorted
    .map(function (entity) {
      const loc = escapeHtml(entity.canonical_url);
      const lastmod = String(entity.updated_at || "").slice(0, 10);
      return (
        "  <url>\n" +
        "    <loc>" +
        loc +
        "</loc>\n" +
        (lastmod ? "    <lastmod>" + lastmod + "</lastmod>\n" : "") +
        "    <changefreq>weekly</changefreq>\n" +
        "    <priority>0.7</priority>\n" +
        "  </url>"
      );
    })
    .join("\n");

  return (
    '<?xml version="1.0" encoding="UTF-8"?>\n' +
    '<!-- P5-E.9F.2 fixture entity sitemap — evidence only, not production sitemap.xml -->\n' +
    '<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">\n' +
    urls +
    "\n</urlset>\n"
  );
}

function resolveFixturePath() {
  const useLegacy = process.argv.includes("--legacy");
  const fixturePath = useLegacy ? LEGACY_FIXTURE : DEFAULT_FIXTURE;
  if (!fs.existsSync(fixturePath)) {
    fail("Fixture file not found: " + fixturePath);
  }
  return fixturePath;
}

function main() {
  const fixturePath = resolveFixturePath();
  const raw = fs.readFileSync(fixturePath, "utf8");
  const data = JSON.parse(raw);
  const entities = data && Array.isArray(data.entities) ? data.entities : null;
  if (!entities || entities.length === 0) {
    fail("No entities in fixture JSON");
  }

  console.log("[build-entity-ssg-fixtures] Reading", entities.length, "entities from", path.relative(ROOT, fixturePath));

  const slugSet = new Set();
  const outputs = [];

  for (let i = 0; i < entities.length; i += 1) {
    const err = validateEntity(entities[i], i);
    if (err) fail(err);
    if (slugSet.has(entities[i].canonical_slug)) {
      fail("duplicate slug: " + entities[i].canonical_slug);
    }
    slugSet.add(entities[i].canonical_slug);

    const slug = entities[i].canonical_slug;
    const outDir = path.join(OUTPUT_BASE, slug);
    const outFile = path.join(outDir, "index.html");
    const html = renderPage(entities[i]);

    fs.mkdirSync(outDir, { recursive: true });
    fs.writeFileSync(outFile, html, "utf8");
    outputs.push(outFile);
    console.log("[build-entity-ssg-fixtures] Wrote", path.relative(ROOT, outFile));
  }

  fs.mkdirSync(path.dirname(NOT_FOUND_OUT), { recursive: true });
  fs.writeFileSync(NOT_FOUND_OUT, renderNotFoundPage(), "utf8");
  console.log("[build-entity-ssg-fixtures] Wrote", path.relative(ROOT, NOT_FOUND_OUT));

  fs.writeFileSync(SITEMAP_OUT, buildSitemapXml(entities), "utf8");
  console.log("[build-entity-ssg-fixtures] Wrote", path.relative(ROOT, SITEMAP_OUT));

  console.log("[build-entity-ssg-fixtures] Done — generated", outputs.length, "entity pages + not-found + sitemap");
}

main();

File: scripts/build-entity-ssg-fixtures.py
#!/usr/bin/env python3
"""P5-E.9F.2 — Python fallback runner for entity SSG generator (same contract as .mjs)."""
import html
import json
import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
FIXTURE = ROOT / "qa" / "fixtures" / "entity-ssg-fixtures.json"
OUTPUT_BASE = ROOT / "wiki" / "post"
SITEMAP_OUT = ROOT / "qa" / "entity-ssg-sitemap.fixture.xml"
NOT_FOUND_OUT = OUTPUT_BASE / "_ssg-not-found" / "index.html"
CANONICAL_BASE = "https://boundlore.com"
GENERATOR_VERSION = "p5-e9f2"
SLUG_RE = re.compile(r"^[a-z0-9]+(?:-[a-z0-9]+)*$")
SLUG_BLOCKLIST = re.compile(r"^(qa|test|p5|fixture|draft|pending|deleted|contribution)")
FORBIDDEN = re.compile(
    r"BLMETA|search_text|search_vector|service_role|SUPABASE_SERVICE_ROLE|"
    r"@[a-z0-9.-]+\.[a-z]{2,}|\b[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\b",
    re.I,
)
HUB_MAP = {
    "creatures": "/wiki/creatures/",
    "items": "/wiki/items/",
    "biomes": "/wiki/biomes/",
    "locations": "/wiki/locations/",
    "resources": "/wiki/resources/",
    "guides": "/wiki/guides/",
    "guilds": "/wiki/guilds/",
    "lore": "/wiki/lore/",
}


def fail(msg: str) -> None:
    print(f"[build-entity-ssg-fixtures] ERROR: {msg}", file=sys.stderr)
    sys.exit(1)


def truncate(text: str, max_len: int) -> str:
    s = (text or "").strip()
    if len(s) <= max_len:
        return s
    return s[: max_len - 1].rstrip() + "…"


def validate_entity(entity: dict, index: int) -> None:
    required = [
        "canonical_slug", "title", "category", "entity_domain", "entity_subtype",
        "excerpt", "body_html_sanitized", "updated_at", "published_at",
        "content_origin", "status", "canonical_url", "breadcrumbs",
    ]
    label = f"entities[{index}]"
    for field in required:
        if not entity.get(field):
            fail(f"{label} missing required field: {field}")
    if entity["status"] != "published":
        fail(f"{label} status must be published")
    if entity["content_origin"] not in {"ssg_fixture", "prototype_fixture"}:
        fail(f"{label} content_origin not allowed")
    slug = entity["canonical_slug"]
    if not SLUG_RE.match(slug) or SLUG_BLOCKLIST.match(slug):
        fail(f"{label} invalid or blocked slug")
    expected = f"{CANONICAL_BASE}/wiki/post/{slug}/"
    if entity["canonical_url"] != expected:
        fail(f"{label} canonical_url mismatch")
    body = entity["body_html_sanitized"]
    if FORBIDDEN.search(body) or "<script" in body.lower():
        fail(f"{label} unsafe body")
    if FORBIDDEN.search(entity["excerpt"]) or FORBIDDEN.search(entity["title"]):
        fail(f"{label} leak in title/excerpt")


def render_page(entity: dict) -> str:
    slug = entity["canonical_slug"]
    title = entity["title"]
    meta_desc = html.escape(entity["excerpt"], quote=True)
    og_desc = html.escape(truncate(entity["excerpt"], 120), quote=True)
    tw_desc = html.escape(truncate(entity["excerpt"], 100), quote=True)
    canonical = html.escape(entity["canonical_url"], quote=True)
    page_title = html.escape(f"{title} – BoundLore Wiki")
    og_title = html.escape(f"{title} – BoundLore Wiki", quote=True)
    image = html.escape(entity.get("image_url") or f"{CANONICAL_BASE}/public/images/social-preview.png", quote=True)
    breadcrumb_html = " &rsaquo; ".join(
        html.escape(c["name"]) if i == len(entity["breadcrumbs"]) - 1
        else f'<a href="{html.escape(c["url"], quote=True)}">{html.escape(c["name"])}</a>'
        for i, c in enumerate(entity["breadcrumbs"])
    )
    badges = "\n".join([
        f'        <span class="bl-ssg-badge">{html.escape("Category: " + entity["category"])}</span>',
        f'        <span class="bl-ssg-badge">{html.escape("Domain: " + entity["entity_domain"])}</span>',
        f'        <span class="bl-ssg-badge">{html.escape("Subtype: " + entity["entity_subtype"])}</span>',
    ])
    related = ""
    if entity.get("relations_summary"):
        links = "\n".join(
            f'          <li><a href="/wiki/post/{html.escape(r.split(":", 1)[-1])}/">{html.escape(r.split(":", 1)[-1].replace("-", " "))}</a></li>'
            for r in entity["relations_summary"]
        )
        related = f"""      <section class="bl-ssg-related" aria-label="Related entries">
        <h2>Related</h2>
        <ul>
{links}
        </ul>
      </section>
"""
    cw = json.dumps({
        "@context": "https://schema.org",
        "@type": "CreativeWork",
        "name": title,
        "description": truncate(entity["excerpt"], 160),
        "url": entity["canonical_url"],
        "datePublished": entity["published_at"],
        "dateModified": entity["updated_at"],
        "image": entity.get("image_url") or f"{CANONICAL_BASE}/public/images/social-preview.png",
        "keywords": f"{entity['category']}, {entity['entity_domain']}, {entity['entity_subtype']}",
        "isPartOf": {"@type": "WebSite", "name": "BoundLore", "url": f"{CANONICAL_BASE}/"},
    }, indent=2)
    bc = json.dumps({
        "@context": "https://schema.org",
        "@type": "BreadcrumbList",
        "itemListElement": [
            {"@type": "ListItem", "position": i + 1, "name": c["name"], "item": c["url"]}
            for i, c in enumerate(entity["breadcrumbs"])
        ],
    }, indent=2)
    return f"""<!-- generated by scripts/build-entity-ssg-fixtures.mjs {GENERATOR_VERSION} -->
<!DOCTYPE html>
<html lang="en" data-bl-ssg="1" data-bl-canonical-slug="{html.escape(slug, quote=True)}" data-bl-ssg-source="fixture-generator">
<head>
  <meta charset="UTF-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <meta name="description" content="{meta_desc}" />
  <meta name="robots" content="noindex, follow" />
  <link rel="canonical" href="{canonical}" />
  <meta property="og:title" content="{og_title}" />
  <meta property="og:description" content="{og_desc}" />
  <meta property="og:type" content="article" />
  <meta property="og:url" content="{canonical}" />
  <meta property="og:image" content="{image}" />
  <meta name="twitter:card" content="summary_large_image" />
  <meta name="twitter:title" content="{og_title}" />
  <meta name="twitter:description" content="{tw_desc}" />
  <meta name="twitter:image" content="{image}" />
  <title>{page_title}</title>
  <link rel="icon" type="image/jpeg" href="/public/images/icon.jpg" />
  <link rel="stylesheet" href="../../../css/style.css" />
  <link rel="preconnect" href="https://fonts.googleapis.com" />
  <link href="https://fonts.googleapis.com/css2?family=Cinzel:wght@400;700;900&family=Inter:wght@300;400;600&display=swap" rel="stylesheet" />
  <script src="/js/error-reporter.js?v=p5-e9c2"></script>
  <script type="application/ld+json">
{cw}
  </script>
  <script type="application/ld+json">
{bc}
  </script>
  <style>
    .bl-ssg-badges {{ display: flex; flex-wrap: wrap; gap: 8px; margin: 12px 0 16px; }}
    .bl-ssg-badge {{ padding: 4px 10px; border-radius: 999px; font-size: 0.75rem; border: 1px solid rgba(255,255,255,0.2); background: rgba(255,255,255,0.06); }}
    .bl-ssg-excerpt {{ font-size: 1.05rem; color: #dde1ef; margin-bottom: 20px; line-height: 1.55; }}
    .bl-ssg-body {{ line-height: 1.6; color: #e8ebf5; }}
    .bl-ssg-related {{ margin-top: 28px; }}
    .bl-ssg-related h2 {{ font-size: 1.1rem; margin-bottom: 10px; }}
    .bl-ssg-related ul {{ padding-left: 1.2rem; }}
    .bl-ssg-hydrate-slot {{ margin-top: 28px; padding: 16px; border: 1px dashed rgba(255,255,255,0.15); border-radius: 8px; color: #9aa8bc; font-size: 0.85rem; }}
  </style>
</head>
<body>
  <nav class="navbar" id="navbar">
    <div class="nav-container">
      <a href="/" class="nav-logo">
        <img src="/public/images/icon.jpg" alt="BoundLore" class="nav-icon" />
        <span class="logo-text">Bound<span class="accent">Lore</span></span>
      </a>
      <ul class="nav-links" id="navLinks">
        <li><a href="/">Home</a></li>
        <li><a href="/wiki/creatures/">Creatures</a></li>
        <li><a href="/wiki/biomes/">Biomes</a></li>
        <li><a href="/wiki/items/">Items</a></li>
        <li><a href="/wiki/browse/">Browse</a></li>
      </ul>
    </div>
  </nav>

  <div class="article-layout bl-post-layout" id="blSsgHydrateRoot" data-bl-ssg-hydrate="1" data-bl-ssg-source="fixture-generator" data-bl-canonical-slug="{html.escape(slug, quote=True)}">
    <article class="article-content bl-post-main">
      <p class="breadcrumb">{breadcrumb_html}</p>

      <h1>{html.escape(title)}</h1>

      <div class="bl-ssg-badges">
{badges}
      </div>

      <p class="bl-ssg-excerpt">{html.escape(entity["excerpt"])}</p>

      <div class="bl-ssg-body">
{entity["body_html_sanitized"]}
      </div>

{related}
      <div class="bl-ssg-hydrate-slot" id="blSsgHydrateRelated" aria-label="CSR hydration placeholder">
        Comments and reactions may load here after hydration. SEO core content is fully static above.
      </div>
    </article>
  </div>

  <footer class="footer"><div class="footer-bottom"><p>&copy; 2026 BoundLore Community &middot; <a href="/wiki/imprint/">Imprint</a> &middot; <a href="/wiki/privacy/">Privacy Policy</a></p></div></footer>
</body>
</html>
"""


def render_not_found() -> str:
    canonical = f"{CANONICAL_BASE}/wiki/post/_ssg-not-found/"
    return f"""<!-- generated by scripts/build-entity-ssg-fixtures.mjs {GENERATOR_VERSION} -->
<!DOCTYPE html>
<html lang="en" data-bl-ssg="1" data-bl-ssg-not-found="1">
<head>
  <meta charset="UTF-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <meta name="description" content="The requested wiki entry was not found." />
  <meta name="robots" content="noindex, nofollow" />
  <link rel="canonical" href="{canonical}" />
  <title>Entry Not Found – BoundLore Wiki</title>
  <link rel="stylesheet" href="../../../css/style.css" />
</head>
<body>
  <main class="article-content" style="max-width:720px;margin:80px auto;padding:24px;">
    <h1>Entry Not Found</h1>
    <p>The requested wiki entry is not available as static SSG output.</p>
    <p><a href="/wiki/browse/">Return to Browse</a></p>
  </main>
</body>
</html>
"""


def build_sitemap(entities: list) -> str:
    urls = []
    for entity in sorted(entities, key=lambda e: e["canonical_slug"]):
        lastmod = entity["updated_at"][:10]
        urls.append(
            f"  <url>\n    <loc>{html.escape(entity['canonical_url'])}</loc>\n"
            f"    <lastmod>{lastmod}</lastmod>\n    <changefreq>weekly</changefreq>\n"
            f"    <priority>0.7</priority>\n  </url>"
        )
    return (
        '<?xml version="1.0" encoding="UTF-8"?>\n'
        "<!-- P5-E.9F.2 fixture entity sitemap — evidence only, not production sitemap.xml -->\n"
        '<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">\n'
        + "\n".join(urls)
        + "\n</urlset>\n"
    )


def main() -> None:
    data = json.loads(FIXTURE.read_text(encoding="utf-8"))
    entities = data.get("entities") or []
    if len(entities) < 1:
        fail("No entities")
    slugs = set()
    for i, entity in enumerate(entities):
        validate_entity(entity, i)
        if entity["canonical_slug"] in slugs:
            fail("duplicate slug")
        slugs.add(entity["canonical_slug"])
        out_dir = OUTPUT_BASE / entity["canonical_slug"]
        out_dir.mkdir(parents=True, exist_ok=True)
        out_file = out_dir / "index.html"
        out_file.write_text(render_page(entity), encoding="utf-8")
        print(f"[build-entity-ssg-fixtures] Wrote {out_file.relative_to(ROOT)}")
    NOT_FOUND_OUT.parent.mkdir(parents=True, exist_ok=True)
    NOT_FOUND_OUT.write_text(render_not_found(), encoding="utf-8")
    print(f"[build-entity-ssg-fixtures] Wrote {NOT_FOUND_OUT.relative_to(ROOT)}")
    SITEMAP_OUT.write_text(build_sitemap(entities), encoding="utf-8")
    print(f"[build-entity-ssg-fixtures] Wrote {SITEMAP_OUT.relative_to(ROOT)}")
    print(f"[build-entity-ssg-fixtures] Done — generated {len(entities)} entity pages + not-found + sitemap")


if __name__ == "__main__":
    main()
